text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,902 @@
+the usual `const fn` rules. This is necessary in order to allow a const block
+(which can access generic parameters) in the function body to invoke methods on the generic parameter.
+
+So, we need some annotation that differentiates a `T: Default` bound from one that gives us the guarantees we're looking for.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Nomenclature and new syntax concepts
+
+### Const trait impls
+
+It is now allowed to prefix a trait name in an impl block with `const`, marking that this `impl`'s type is now allowed to
+have methods of this `impl`'s trait to be called in const contexts (if all where bounds hold, like ususal, but more on this later).
+
+An example looks as follows:
+
+```rust
+impl const Trait for Type {}
+```
+
+Such impls require that the trait is a `const trait`.
+
+All method bodies in a const trait impl are [const contexts](https://doc.rust-lang.org/reference/const_eval.html#const-context).
+
+### Const traits
+
+Traits need to opt-in to being allowed to have const trait impls. Thus you need to declare your traits by prefixing the `trait` keyword with `const`:
+
+```rust
+const trait Trait {}
+```
+
+This in turn checks all methods' default bodies as if they were `const fn`, making them callable in const contexts.
+Impls can now rely on the default methods being const, too, and don't need to override them with a const body.
+
+We may add an attribute later to allow you to mark individual trait methods as not-const so that when creating a const trait, one can
+add (defaulted or not) methods that cannot be used in const contexts.
+
+It is possible to split up a trait into the const an non-const parts as discussed [here](#cant-have-const-methods-and-nonconst-methods-on-the-same-trait).
+
+All default method bodies of const trait declarations are [const contexts](https://doc.rust-lang.org/reference/const_eval.html#const-context).
+
+Note that on nightly the syntax is
+
+```rust
+#[const_trait]
+trait Trait {}
+```
+
+and a result of this RFC would be that we would remove the attribute and add the `const trait` syntax.
+
+### Const trait bounds
+
+Any item that can have trait bounds can also have `const Trait` bounds.
+
+Examples:
+
+* `T: const Trait`, requiring any type that `T` is instantiated with to have a const trait impl.
+* `dyn const Trait`, requiring any type that is unsized to this dyn trait to have a const trait impl.
+    * These are not part of this RFC because they require `const fn` function pointers. See [the Future Possibilities section](#future-possibilities).
+* `impl const Trait` (in all positions).
+    * These are not part of this RFC because they require `const fn` function pointers. See [the Future Possibilities section](#future-possibilities).
+* `trait Foo: const Bar {}`, requiring every type that has an impl for `Foo` (even a non-const one), to also have a const trait impl for `Bar`.
```

Comment:
With this RFC in its current state, `Copy` unfortunately does not imply `const Clone`. I don’t know how likely this is to cause problems in practice, or how difficult it would be to fix, but it probably deserves a mention at least.

Reply:
Probably on the same order of difficulty as fixing the fact that `impl Copy for T` and `impl Clone for T` may have diverging behaviors, i.e. the latter could have observable side-effects. Would prefer if we didn't fix it for now though.